pkg/tests/stamp_test.py

```diff
@@ -13,17 +13,52 @@
 # limitations under the License.
 """Test time stamping in pkg_tar"""
 
+import datetime
 import tarfile
 import time
 import unittest
+import zipfile
 
 from bazel_tools.tools.python.runfiles import runfiles
 
 # keep in sync with archive.py
 PORTABLE_MTIME = 946684800  # 2000-01-01 00:00:00.000 UTC
 
-class PkgTarTest(unittest.TestCase):
-  """Testing for pkg_tar rule."""
+
+class StampTest(unittest.TestCase):
+  """Test for time stamps in packages."""
+
+  target_mtime = int(time.time())
+  zip_epoch_dt = datetime.datetime(1980, 1, 1, 0, 0, 0, tzinfo=datetime.timezone.utc)
+  ZIP_EPOCH = int(zip_epoch_dt.timestamp())
+  ALLOWED_DELTA_FROM_NOW = 10000  # seconds
+
+  def check_mtime(self, mtime, file_path, file_name):
+    """Checks that a time stamp is reasonable.
+
+    This checks that a timestamp is not 0 or any of the well known EPOCH values,
+    and that it is within some delta from the current time.
+
+    Args:
+      mtime: timestamp in seconts
+      file_path: path to archive name
+      file_name: file within archive
+    """
+    if mtime == 0:
+       self.fail('Archive %s contains file %s with mtime == 0' % (
+           file_path, file_name))
+    if mtime == PORTABLE_MTIME:
+       self.fail('Archive %s contains file %s with portable mtime' % (
+           file_path, file_name))
+    if mtime == StampTest.ZIP_EPOCH:
+       self.fail('Archive %s contains file %s with ZIP epoch' % (
+           file_path, file_name))
+    if ((mtime < self.target_mtime - StampTest.ALLOWED_DELTA_FROM_NOW)
+        or (mtime > self.target_mtime + StampTest.ALLOWED_DELTA_FROM_NOW)):
+       self.fail(
+           'Archive %s contains file %s with mtime:%d, expected:%d +/- %d' % (
+               file_path, file_name, mtime, self.target_mtime,
+               StampTest.ALLOWED_DELTA_FROM_NOW))
```

Comment:
Consider using asserts instead of this conditional logic -- `self.assertNotEqual` and `self.assertAlmostEqual` come to mind.

[Documentation](https://docs.python.org/3/library/unittest.html).

Reply:
AlmostEqual is probably too precise.  If our tests get big enough, we could build the files minutes before using them in a test, so the timestamp could as far off of "now()" as the total length of test time. I chose this as a reasonable hedge against variability vs time to build super robust tests if we don't need the extra robustness.

Reply:
I added these comments into the code.